### DISCUSSION TITLE
v2.0.0-beta.0 - Open Comment And Feedback Discussion

### DISCUSSION THREAD

**Discussion by sreeise** [sreeise/graph-rs-sdk#454] [General]
- Complete rewrite of SDK Client for the Microsoft Identity Platform
- In Memory Token Cache
- Automatic Token Refresh
- Interactive Auth Using WebView

**Full Changelog**: https://github.com/sreeise/graph-rs-sdk/compare/v1.1.3...v2.0.0-beta.0

<hr /><em>This discussion was created from the release <a href='https://github.com/sreeise/graph-rs-sdk/releases/tag/v2.0.0-beta.0'>v2.0.0-beta.0</a>.</em>

**Comment by sreeise** [comment]
Feel free to post comments about your usage of pre-release 2.0.0-beta.0. Please post if you find any issues or have questions and provide feedback on what can be improved and how your using the SDK.

Anyone can post about the pre-release. Questions about how to use the new features and other related questions are welcome as well.

**Comment by samvv** [comment]
I really like it. I have one minor question which I posted in #464 but aside from that the beta works really well. Thank you for creating it!

**Comment by ghost** [comment]
It works really well, but I have a small nitpick with DeviceCodeAuthorization, namely that closing the window after authenticating causes my entire application to stop.

This is how I am authenticating (yes it is cursed but I'm not good at oauth :P)
```rs
pub fn authenticate(
    tenant_id: &str,
    client_id: &str,
    scope: &[&str],
) -> Result<GraphClient, Whatever> {
    let device_code_builder = PublicClientApplicationBuilder::new(client_id)
        .with_scope(scope)
        .with_tenant(tenant_id)
        .with_device_code_executor()
        .with_interactive_authentication();

    let (auth_response, mut auth) = match device_code_builder {
        Ok((resp, auth)) => (resp, auth),
        Err(e) => {
            whatever!("failed to obtain device code: {e}");
        }
    };

    tracing::info!(
        "PLEASE FILL IN THE DEVICE CODE: {}",
        auth_response.user_code
    );

    let client = match auth.interactive_webview_authentication(WebViewOptions::default()) {
        Ok(c) => c,
        Err(e) => {
            whatever!("failed to authenticate with device code: {e}");
        }
    };

    Ok(GraphClient::from(&client))
}
```
and this works well. However, after this finishes, the window used to input the code still remains, and if I close it, I get these logs and then the application stops.
```
2024-03-20T10:00:21.619940Z DEBUG graph_rs_sdk::interactive_auth: window close requested by user
2024-03-20T10:00:22.121913Z DEBUG graph_rs_sdk::interactive_auth: window destroyed
```
So I mainly just want a way to be rid of the webview window after authentication 😊

Thanks a lot for this crate 🥰

**Reply by sreeise** [reply]
Hey, thanks for the feedback. Yeah this one was annoying me as well. This is something that happens by default using wry and tao crates for the WebView. It should be fixed once 2.0 comes out. 

Basically it will run the WebView on a separate thread that is not the main thread and so when it shutsdown it won't take the whole application with it. If im not mistaken, the beta the auth code credential does have this implemented to run in a different thread but not the device code credential. I will double check everything for the 2.0 release to make sure all of the credentials work without killing the main thread.

Glad you like the crate as well. Thanks.

**Reply by ghost** [reply]
Great to hear! Can't wait for 2.0 🎉

**Comment by sreeise** [comment]
2.0.0 has been released 🎉
